// memory.proto

syntax = "proto3";
import "common.proto";
package memory;


option go_package = "authentication_service/internal/proto/memory";

service MemoryService {
  rpc CreateMemory (CreateMemoryRequest) returns (CreateMemoryResponse);
  rpc GetMemoryById (GetMemoryByIdRequest) returns (GetMemoryByIdResponse);
  rpc GetMemoriesByDocId (GetMemoriesByDocIdRequest) returns (GetMemoriesByDocIdResponse);
  rpc GetMemoriesByMemoryTitleSearch (GetMemoriesByMemoryTitleSearchRequest) returns (GetMemoriesByMemoryTitleSearchResponse);
  rpc UpdateMemory (UpdateMemoryRequest) returns (UpdateMemoryResponse);
  rpc DeleteMemory (DeleteMemoryRequest) returns (DeleteMemoryResponse);
}


// MemoryService
message CreateMemoryRequest {
  uint64 user_id = 1;
  uint64 doc_id = 2;
  string memory_title = 3;
  string memory_content = 4;
  bool is_public = 5;
}

message CreateMemoryResponse {
}

message GetMemoryByIdRequest {
  uint64 user_id = 1;
  uint64 memory_id = 2;
}

message GetMemoryByIdResponse {
  DBMemory memory = 1;
}

message GetMemoriesByDocIdRequest {
  uint64 user_id = 1;
  uint64 doc_id = 2;
  uint32 page_number = 3;
  uint32 page_size = 4;
  common.ORDER_BY_FIELD order_by_field = 5;
  common.ORDER_BY_DIRECTION order_by_direction = 6;
}

message GetMemoriesByDocIdResponse {
  repeated DBMemory memories = 1;
}

message GetMemoriesByMemoryTitleSearchRequest {
  uint64 user_id = 1;
  string search_query = 2;
  uint32 page_number = 3;
  uint32 page_size = 4;
  common.ORDER_BY_FIELD order_by_field = 5;
  common.ORDER_BY_DIRECTION order_by_direction = 6;
}

message GetMemoriesByMemoryTitleSearchResponse {
  repeated DBMemory memories = 1;
}

message UpdateMemoryRequest {
  uint64 user_id = 1;
  uint64 memory_id = 2;
  string memory_title = 3;
  string memory_content = 4;
}

message UpdateMemoryResponse {
}

message DeleteMemoryRequest {
  uint64 user_id = 1;
  uint64 memory_id = 2;
}

message DeleteMemoryResponse {
}


// DB Models

message DBMemory {
  uint64 memory_id = 1;
  uint64 doc_id = 2;
  string memory_title = 3;
  string memory_content = 4;
  uint64 created_time = 5;
  uint64 updated_time = 6;
}